main_test.go

```diff
@@ -0,0 +1,224 @@
+		Max:        max,
+		Deployment: kubernetesDeploymentName,
+	}
+}
+
+type MockSQS struct {
+	QueueAttributes *sqs.GetQueueAttributesOutput
+}
+
+func (m *MockSQS) GetQueueAttributes(*sqs.GetQueueAttributesInput) (*sqs.GetQueueAttributesOutput, error) {
+	return m.QueueAttributes, nil
+}
+
+func (m *MockSQS) SetQueueAttributes(input *sqs.SetQueueAttributesInput) (*sqs.SetQueueAttributesOutput, error) {
+	m.QueueAttributes = &sqs.GetQueueAttributesOutput{
+		Attributes: input.Attributes,
+	}
+	return &sqs.SetQueueAttributesOutput{}, nil
+}
+
+func NewMockSqsClient() *mainsqs.SqsClient {
+	Attributes := make(map[string]*string)
+	Attributes["ApproximateNumberOfMessages"] = aws.String("50")
+
+	return &mainsqs.SqsClient{
+		Client: &MockSQS{
+			QueueAttributes: &sqs.GetQueueAttributesOutput{
+				Attributes: Attributes,
+			},
+		},
+		QueueUrl: "sqs.test.com",
+	}
+}
+
+func TestRunMaxMin(t *testing.T) {
+	// override default vars for testing
+	*pollInterval = 1 * time.Second
+	*scaleDownCoolPeriod = 1 * time.Second
+	*scaleUpCoolPeriod = 1 * time.Second
+	*scaleUpMessages = 100
+	*scaleDownMessages = 10
+	*maxPods = 5
+	*minPods = 1
+	*awsRegion = "us-east-1"
+
+	*sqsQueueUrl = "sqs.test.com"
+	*kubernetesDeploymentName = "test"
+	*kubernetesNamespace = "test"
+
+	p := NewMockPodAutoScaler(*kubernetesDeploymentName, *kubernetesNamespace, *maxPods, *minPods)
+	s := NewMockSqsClient()
+
+	go Run(p, s)
+
+	Attributes := make(map[string]*string)
+	Attributes["ApproximateNumberOfMessages"] = aws.String("100")
+	input := &sqs.SetQueueAttributesInput{
+		Attributes: Attributes,
+	}
+	s.Client.SetQueueAttributes(input)
+
+	time.Sleep(10 * time.Second)
+	deployment, _ := p.Client.Deployments("test").Get("test")
+	assert.Equal(t, int32(*maxPods), deployment.Spec.Replicas)
+
+	Attributes = make(map[string]*string)
+	Attributes["ApproximateNumberOfMessages"] = aws.String("10")
+	input = &sqs.SetQueueAttributesInput{
+		Attributes: Attributes,
+	}
+	s.Client.SetQueueAttributes(input)
+
+	time.Sleep(10 * time.Second)
+	deployment, _ = p.Client.Deployments("test").Get("test")
+	assert.Equal(t, int32(*minPods), deployment.Spec.Replicas)
+}
+
+func TestRunScaleUpCoolDown(t *testing.T) {
+	// override default vars for testing
+	*pollInterval = 5 * time.Second
+	*scaleDownCoolPeriod = 10 * time.Second
+	*scaleUpCoolPeriod = 10 * time.Second
+	*scaleUpMessages = 100
+	*scaleDownMessages = 10
+	*maxPods = 5
+	*minPods = 1
+	*awsRegion = "us-east-1"
+
+	*sqsQueueUrl = "sqs.test.com"
+	*kubernetesDeploymentName = "test"
+
+	p := NewMockPodAutoScaler(*kubernetesDeploymentName, *kubernetesNamespace, *maxPods, *minPods)
+	s := NewMockSqsClient()
+
+	go Run(p, s)
+
+	Attributes := make(map[string]*string)
+	Attributes["ApproximateNumberOfMessages"] = aws.String("100")
+	input := &sqs.SetQueueAttributesInput{
+		Attributes: Attributes,
+	}
+	s.Client.SetQueueAttributes(input)
+
+	// with a 15 second wait and a 10 second cool down, we should expect only 1 pod to scale up
```

Comment:
is that true?
t=0 poll -> scale up
t=5 poll -> cool down noop
t=10.01 poll -> scale up (cool down has expired)